### PR TITLE
fix #157, Add  X-Zerocracy-Host listening

### LINKED ISSUE #157: the `X-Zerocracy-Host` header must be taken into account
When we do `upload()` and `download()` in the `BazaRb`, we should listen to the `X-Zerocracy-Host` header returned back to us by the server. The value in the header is the hostname of the server where we should send our next request. The original name provided in the `@host` must be discarded and new name must be used.

This is necessary for multi-server setup, when we start uploading to one server and don't want to continue to another server. We need to stay with the same server all the time.

### PR DESCRIPTION
**Pull request by paveltrty**
## Summary by CodeRabbit

* **Bug Fixes**
  * Safer dynamic host handling: client now validates and applies server-suggested host changes during downloads/uploads (including mid-transfer), updates the active host safely, logs switches, and avoids unnecessary changes.

* **Tests**
  * Added tests covering host redirection and mid-operation host switching for downloads/uploads (including range/chunked flows) and validation edge cases.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: test/test_baza-rb.rb [hunk 1]
```diff
@@ -748,6 +749,288 @@ def test_durable_load_from_sinatra
     end
   end
 
+  def test_download_updates_host_from_x_zerocracy_host_header
+    WebMock.disable_net_connect!
+    Dir.mktmpdir do |dir|
+      file = File.join(dir, 'test.txt')
+      initial_host = 'example.org'
```
Comment (yegor256) [inline test/test_baza-rb.rb]: @paveltrty see https://www.yegor256.com/2015/01/12/compound-name-is-code-smell.html

File: lib/baza-rb.rb [hunk 2]
```diff
@@ -579,6 +580,38 @@ def csrf
 
   private
 
+  # Update host from X-Zerocracy-Host header if present.
+  #
+  # @param [Typhoeus::Response] ret The HTTP response containing headers
+  # @param [Iri] uri The current URI object to update
+  # @return [Iri] The updated URI object (or original if no valid header present)
+  # @note Invalid hostnames are logged as warnings and ignored
+  def update_host_from_response(ret, uri)
```
Comment (yegor256) [inline lib/baza-rb.rb]: @paveltrty see https://www.yegor256.com/2018/08/22/builders-and-manipulators.html

File: lib/baza-rb.rb [hunk 3]
```diff
@@ -579,6 +580,38 @@ def csrf
...
+    end
+    uri
+  end
+
+  # Validate hostname format according to RFC 1123.
+  #
+  # @param [String] hostname The hostname to validate
+  # @return [Boolean] True if valid, false otherwise
+  def valid_hostname?(name)
+    return false unless name.is_a?(String)
```
Comment (yegor256) [inline lib/baza-rb.rb]: @paveltrty how "non string" may happen in this private method? see https://github.com/yegor256/at?tab=readme-ov-file#3-31

File: lib/baza-rb.rb [hunk 4]
```diff
@@ -579,6 +580,38 @@ def csrf
...
+    name = name.strip
+    return false if name.empty? || name.bytesize > 253
+    !!name.match?(/\A[a-zA-Z0-9]([a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(\.[a-zA-Z0-9]([a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*\.?\z/)
```
Comment (yegor256) [inline lib/baza-rb.rb]: @paveltrty what `!!` is for?

**Comment by paveltrty** [comment]
@yegor256, fix CI

**Comment by paveltrty** [comment]
@yegor256 check please

**Comment by yegor256** [comment]
@paveltrty thanks!
